File: LeetCode/Medium/Encode Decode TinyURL/Java/EncodeDecodeTinyURL.java


//In this method we use the inbuilt method of java hashcode
/*
The hash code for a String object is computed(using int arithmetic) as −

s[0]*31^{(n - 1)} + s[1]*31^{(n - 2)} + ... + s[n - 1]s[0]∗31 
(n−1)
 +s[1]∗31 
(n−2)
 +...+s[n−1] , where s[i] is the ith character of the string, n is the length of the string.
 */

public class Codec {
	Map<Integer, String> map = new HashMap<>();
	public String encode(String longUrl) {
		map.put(longUrl.hashCode(), longUrl);
		return "https://tinyurl.com/" + longUrl.hashCode();
	}

	public String decode(String shortUrl) {
		return map.get(Integer.parseInt(shortUrl.replace("https://tinyurl.com/","")));
	}
}

File: LeetCode/Medium/Encode Decode TinyURL/Go/encode_decode_tiny_url.go
package main

type Codec struct {
    hash_map map[int]string
    i int
}


func Constructor() Codec {
    return Codec{hash_map:make(map[int]string), i:0 }
}

// Encodes a URL to a shortened URL.
func (this *Codec) encode(longUrl string) string {
	this.i++
    this.hash_map[this.i] = longUrl
    return "https://tinyurl.com/" + strconv.Itoa(this.i)
}

// Decodes a shortened URL to its original URL.
func (this *Codec) decode(shortUrl string) string {
    v := int(shortUrl[len(shortUrl)-1]-'0')
    return this.hash_map[v]
}

func main(){}